cmake/kokkos_enable_options.cmake

```
@@ -102,6 +102,11 @@ kokkos_enable_option(
   "Whether multiple kernels are instantiated at compile time - improve performance but increase compile time"
 )
 kokkos_enable_option(IMPL_HIP_MALLOC_ASYNC ${KOKKOS_ENABLE_HIP} "Whether to enable hipMallocAsync")
+if(INTERNAL_ROCM_VERSION VERSION_EQUAL 7.1.0)
+  set(KOKKOS_ENABLE_IMPL_HIP_MALLOC_ASYNC OFF)
+  message(WARNING "Disabling IMPL_HIP_MALLOC_ASYNC when using ROCm 7.1.0")
+endif()
```

```suggestion
if(INTERNAL_ROCM_VERSION VERSION_EQUAL 7.1.0)
  set(HIP_MALLOC_ASYNC_DEFAULT OFF)
else()
  set(HIP_MALLOC_ASYNC_DEFAULT ON)
endif()
kokkos_enable_option(IMPL_HIP_MALLOC_ASYNC ${HIP_MALLOC_ASYNC_DEFAULT} "Whether to enable hipMallocAsync")
if(INTERNAL_ROCM_VERSION VERSION_EQUAL 7.1.0 AND Kokkos_ENABLE_IMPL_HIP_MALLOC_ASYNC)
  message(WARNING "Using Kokkos_ENABLE_IMPL_HIP_MALLOC_ASYNC using ROCm 7.1.0 is known to be problematic")
endif()
```
so that we keep `Kokkos_ENABLE_IMPL_HIP_MALLOC_ASYNC` and `KOKKOS_ENABLE_IMPL_HIP_MALLOC_ASYNC` in sync.

---

Why do you introduce `HIP_MALLOC_ASYNC`? You are not using it

> so that we keep Kokkos_ENABLE_IMPL_HIP_MALLOC_ASYNC and KOKKOS_ENABLE_IMPL_HIP_MALLOC_ASYNC in sync.

I want to overwrite the value set by the user. If they enable async, they will get random errors.

---

> Why do you introduce `HIP_MALLOC_ASYNC`? You are not using it

I fixed it above so that we use `HIP_MALLOC_ASYNC_DEFAULT` as default for the option.

> I want to overwrite the value set by the user. If they enable async, they will get random errors.

You are overwriting `KOKKOS_ENABLE_IMPL_HIP_MALLOC_ASYNC` (non-cached) but not the actual option `Kokkos_ENABLE_IMPL_HIP_MALLOC_ASYNC` (cached). If we want to force it, we should error out instead of warn.